MAINT: Add optional-dependencies to pyproject.toml

nialov opened the pull request:
## What

The specification for `optional-dependencies` is part of PEP 621
<https://peps.python.org/pep-0621/>. See
<https://packaging.python.org/en/latest/specifications/pyproject-toml/#pyproject-toml-spec>
for latest specification.

I added the dependencies to `optional-dependencies` based on
`./requirements-dev.txt`, `./environment.yml` and #2990.

## Why?

This would be useful for defining the optional dependencies when packaging
`geopandas` for e.g. `nixpkgs`
(<https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/geopandas/default.nix>)
as it is better for them to be defined here in the main repository than
downstream. As there is already code in `./geopandas/` that use these optional
dependencies, I believe they should be documented somewhere.

## Misc

I did not update `requirements-dev.txt` as I do not know if it is currently
up-to-date (<https://github.com/geopandas/geopandas/pull/2438#issuecomment-1205268708>)
or being maintained.

Marked as draft as no feedback yet given from maintainers. Hopefully this sounds good!

m-richards commented on pyproject.toml:
```diff
@@ -28,6 +28,28 @@ dependencies = [
+    "geoplot",
+    "folium",
+]
+postgis = [
```
```suggestion
postgis-write = [
```

m-richards commented on pyproject.toml:
```diff
@@ -28,6 +28,28 @@ dependencies = [
 ]
 
+[project.optional-dependencies]
+geodatabase = [
```
I think we should try and align with pandas which has options for `postgresql`, `mysql`, `sql-other`, hence postgis (we could have a generic `sql` but it would only have one package in it).
```suggestion
postgis = [
```

m-richards commented on pyproject.toml:
```diff
@@ -28,6 +28,28 @@ dependencies = [
+    "matplotlib>=3.5.0",
+    "mapclassify",
+    "cartopy",
+    "geoplot",
```
```suggestion
    "xyzservices",
```
geoplot is no longer maintained and should be removed, and I think it makes sense to include xyzservices here.  

I also think cartopy probably should not be here since we don't actually have any dependence on it - it's just part of the ecosystem.

m-richards reviewed (commented):
Thanks for adding this @nialov, seems like a good contribution to me. I've added some small suggestions.

If you wouldn't mind adding pyarrow to the requirements-dev.txt as well I think that's a good idea. I don't think any of the maintainers actually use this generally, and our docs recommend conda, but having something in a pip friendly format can still be useful for anyone not using conda.

nialov commented:
Took a look at some other packages and I am actually wondering if it would be simpler to just define two optional dependency groups: `all` and `dev`. `all` includes all dependencies except the development ones. `dev` contains `pytest` and others needed for testing and development.`all` is used by well-known packages, such as `pandas` and `typer`, to indicate all optional dependencies. 

If `all` could be defined by pointing to the already defined groups (`postgis`, `plot`, etc.), that would be the best way to define `all`. But that is not possible without introducing a possible circular dependency issue. E.g. `pandas` defines `all` by redefining all the dependencies of all other groups in `all` (<https://github.com/pandas-dev/pandas/blob/56b5979f136e72ce78e5221392739481c025d5e7/pyproject.toml#L88>). This is not nice to maintain.

The grouping feature is not apparent to users and writing e.g. `pip install geopandas[postgis,postgis-write,plot]` is demanding a lot from any casual installer of `geopandas`. I would also wager that either a) an user wants to minimize dependencies, in which case they can add the optional dependencies as they run into errors as they run `geopandas` code or b) just want all functionality of `geopandas` quickly. For a) the groupings are not much better than just manually installing the required packages. For b) the `all` group gives a simple way to install all optionals.

martinfleis commented:
I just want to say that I believe it is nice to have optional dependencies defined in some way in pyproject. However, I don't think I've ever used it a as a conda/mamba user myself so don't really care if it is split in groups of just `all` and `dev`. The latter seems more maintainable.

m-richards commented:
I also don't have a strong leaning on all & dev vs split out. I use pip rather than conda for work, but I don't tend to use extras, I'll just install the base package. Even if I know and extra exists I usually have to double check what it's called  - and the most reliable way of doing that is looking at the pyproject.toml, at which point I no longer need the extra because I can see the packages underneath it.

nialov commented:
Updated the pull request to use the two groups discussed above, `all` and `dev`. Also updated documentation to inform users and developers of these options. One change, that I do not know if is appropriate, is the use of `psycopg2-binary` instead of `psycopg2`. Using this package is easier as it contains binary components of `psycopg2` directly (<https://github.com/psycopg/psycopg2#installation>).

m-richards commented on doc/source/getting_started/install.rst:
```diff
@@ -85,6 +85,10 @@ as well::
 
     pip install geopandas
 
+Optional runtime dependencies can also be installed at once::
```
```suggestion
Optional runtime dependencies can also be installed all at once::
```

I'd also suggest putting this after the warning text just below this - as the warning is for required deps.

nialov replied:
Thanks, done!

m-richards reviewed (approved):
I think switching to psycopg2-binary is fine, one small comment. Thanks @nialov!

nialov commented:
Ready for (full) review, sorry for slight delay!

nialov commented on doc/source/getting_started/install.rst:
```diff
@@ -111,6 +115,11 @@ You may install the latest development version by cloning the
+Development dependencies can be installed using the dev optional
+dependency group::
+
+    pip install .[dev]
```
E.g. this command worked on my machine when creating a virtual environment:

~~~python
git clone git@github.com:nialov/geopandas.git
cd geopandas
git checkout 2990-add-optional-dependencies
python3 -m venv .venv
.venv/bin/python3 -m pip install .[dev] --dry-run
~~~

Could it be an older `pip` version that is only reading `setup.py`? I am not an expert on `pip`, so could not say what else could be the reason.

martinfleis replied:
Yes please. That is also what pandas is apparently [doing](https://pandas.pydata.org/docs/getting_started/install.html#installing-from-pypi).

nialov replied:
Done 👍

martinfleis reviewed (approved):
Thanks!
